keps/sig-instrumentation/4200-healthcheck-metrics/README.md

```diff
@@ -0,0 +1,297 @@
+of these inputs.
+
+## Motivation
+
+Healthchecking data is currently surfaced in unstructured format and is scraped by monitoring
+agents (as well as the kubelet), which must be configured to interpret the health data and
+to act upon it. This process does not lend itself readily to the creation of availability SLOs,
+since we basically require an outside agent to parse the health endpoint and convert this signal
+into an SLI. 
+
+### Goals
+
+- Create a uniform interface by which we can consume health checking information
+- Allow availability SLIs to be created without a specialized monitoring agent
+
+
+### Non-Goals
+
+- Creation of SLOs are out of scope. This KEP specifically targets the creation of signals which can be used as SLIs.
+
+## Proposal
+
+We are proposing to add a new endpoint on Kubernetes components `/health/metrics` which returns
+health check data in prometheus format.
+
+
+### Risks and Mitigations
+
+This is a separate endpoint, it does not have to be used. The risk of adding metrics is generally cardinality, but in this
+case we are proposing known dimensions to the metrics, specifically:
+
+1. status - one of `Success`, `Error`, `Pending`
+1. type - one of `livez`, `readyz`, `healthz`
+1. name - the known name of the health check. AFAIK, these are all static strings in the Kubernetes codebase, therefore bounded in cardinality
+
+
+## Design Details
+
+
+### Test Plan
+
+
+##### Unit tests
+
+- [ ] ensure that healthcheck states are reset for gauges on write
+- [ ] ensure that counters properly retain state of all seen healthchecks
+
+##### Integration tests
+
+- [ ] ensure existence of healthcheck endpoint
+
+##### e2e tests
+
+
+- [ ] ensure existence of healthcheck endpoint
+
+
+### Graduation Criteria
+
+
+## Production Readiness Review Questionnaire
+
+
+### Feature Enablement and Rollback
+
+We can feature gate the api and the registry for this metrics endpoint, such that it is disabled by default and toggleable only by feature-gate. Given the lazy
+nature of the Kubernetes metric framework, this will ensure that we do not encounter any memory leaks.
+
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+Via feature gate.
+
+
+###### Does enabling the feature change any default behavior?
+
+Yes, it exposes an endpoint which can then be scraped.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes, but it will remove the endpoint.
+
+### Monitoring Requirements
+
+I am proposing a metric/series of metrics here.
+
+### Scalability
+
+We already hit the readyz/healthz/livez endpoints of our control-plane components frequently, this KEP only 
+adds instrumentation of these endpoints' results.
+
+###### Will enabling / using this feature result in any new API calls?
+
+Yes, we are proposing that this health metrics are surfaced in each component under `/health/metrics` which 
```

Comment:
```suggestion
Yes, we are proposing that this health metrics are surfaced in each component under `/metrics/health` which 
```

I think most of the time, the pattern has been `/metrics/foo`, rather than `/foo/metrics`.  At least for the examples I can think of...

`/metrics/resource` `/metrics/cadvisor` `/metrics/probes` from https://kubernetes.io/docs/concepts/cluster-administration/system-metrics/#metrics-in-kubernetes

Reply:
Agreed, i wonder if we should document that in api-conventions?  Would save time.

Reply:
Yeah sounds good to me.